components/ws-daemon/pkg/controller/workspace_controller.go

```diff
@@ -219,7 +221,29 @@ func (wsc *WorkspaceController) handleWorkspaceRunning(ctx context.Context, ws *
 	span, ctx := opentracing.StartSpanFromContext(ctx, "handleWorkspaceRunning")
 	defer tracing.FinishSpan(span, &err)
 
-	return ctrl.Result{}, wsc.operations.SetupWorkspace(ctx, ws.Name)
+	if ws.Status.ImageInfo == nil {
+		id, err := wsc.runtime.WaitForContainer(ctx, ws.Name)
+		if err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed to wait for container: %w", err)
+		}
+		info, err := wsc.runtime.GetContainerImageInfo(ctx, id)
+		if err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed to get container image info: %w", err)
+		}
+
+		err = retry.RetryOnConflict(retryParams, func() error {
+			if err := wsc.Get(ctx, req.NamespacedName, ws); err != nil {
+				return err
+			}
+			ws.Status.ImageInfo = info
+			return wsc.Status().Update(ctx, ws)
+		})
+		if err != nil {
+			glog.WithFields(ws.OWI()).WithField("workspace", req.NamespacedName).WithField("phase", ws.Status.Phase).Errorf("failed to update workspace with image info: %v", err)
+		}
+		wsc.operations.SetupWorkspace(ctx, ws.Name, info)
```

Comment:
The double `SetupWorkspace` call made me double-check the code. It's fine, but probably not expected to see a function called "setupXYZ" called twice - and honestly, it's a misnomer from the start.

What about setting a variable `imageInfo` and handle the last call everything?